docs/config-files.md

```diff
@@ -0,0 +1,228 @@
+  mod1/
+    package.json
+    src/index.js
+  mod2/
+    package.json
+    src/index.js
+```
+the config will now be entirely ignored, because it is across a package boundary.
+
+One alternative would be to create a `.babelrc` in each sub-package that uses ["extends"](options.md#extends) as
+```json
+{ "extends": "../../.babelrc" }
+```
+Unfortunately, this approach can be a bit repetitive, and depending on how Babel is being used,
+could require setting ["babelrcRoots"](options.md#babelrcRoots).
+
+Given that, it may be more desirable to rename the `.babelrc` to be a
+[project-wide "babel.config.js"](#project-wide-configuration). As mentioned in the project-wide
+section above, this may then require explicitly setting ["configFile"](options.md#configFile)
+since Babel will not find the config file if the working directory isn't correct.
+
+
+## Config Format
+
+The format of individual config files themselves separates into JS files vs [JSON5](https://json5.org/) files.
+
+### JSON5
+
+Any file that isn't a `.js` file will be parsed as JSON5 and should contain an object matching
+the [options](options.md) format that Babel accepts.
+
+### JavaScript
+
+Any `.js` file will be `require()`ed and should export either a configuration object, or a function
+that will return a configuration object when called. The main benefit being that users can include
+JS logic to build up their config structures, potentially allowing config logic to be shared
+more easily. `.js` files can be used as [project-wide configuration](#project-wide-configuration) or
+via `.babelrc.js` files for [file-relative configuration](#file-relative-configuration).
+
+Function-returning configs are given a few special powers because they can access an API exposed
+by Babel itself. See [Config Function API](#config-function-api) for more information.
+
+## Config Function API
+
+JS config files may export a function that will be passed config function API:
+
+```js
+module.exports = function(api) {
+  return {};
+}
+```
+
+The `api` object exposes everthing Babel itself exposes from its index module, along with
+config-file specific APIs:
+
+### `api.version`
+
+Type: `string`<br />
+
+The version string for the Babel version that is loading the config file.
+
+### `api.cache`
+
+JS configs are great because they can compute a config on the fly, but the downside
+there is that it makes caching harder. Babel wants to avoid re-executing the
+config function every time a file is compiled, because then it would also need to
+re-execute any plugin and preset functions referenced in that config.
+
+The avoid this, Babel expects users of config functions to tell it how to manage
+caching within a config file.
+
+* `api.cache.forever()` - Permacache the computed config and never call the function again.
+* `api.cache.never()` - Do not cache this config, and re-execute the function every time.
+* `api.cache.using(() => process.env.NODE_ENV)` - Cache based on the value of `NODE_ENV`.
```

Comment:
So any of the cache validation functions receive any argument? That lead them to work with global variables and... Nothing else because scope will not be retained between calls, or will it be? What if you want to validate the cache based on the files that have changed or something like that? Why not provide an argument to this functions?

Reply:
What argument would it provide exactly? The primary usecase for these functions is to check for global flags that should affect config loading, so there isn't generally any state to pass in.

> What if you want to validate the cache based on the files that have changed or something like that?

You can potentially have the cache function do `api.cache.invalidate(() => fs.statSync("somefile").mtime)` to invalidate the config cache when that file changes.

Reply:
I thought that babel (watch for example) would keep some kind of report if what has changed and what it is intended to do.

Reply:
Not currently.